engine/utils/vec/Vec4.h

```diff
@@ -0,0 +1,411 @@
+//
+//  Copyright (c) 2022 Jonathan Moallem (@J-Mo63) & Aryeh Zinn (@Raelr)
+//
+//  This code is released under an unmodified zlib license.
+//  For conditions of distribution and use, please see:
+//      https://opensource.org/licenses/Zlib
+
+#ifndef SIEGE_ENGINE_VEC4_H
+#define SIEGE_ENGINE_VEC4_H
+
+#include "Vec3.h"
+
+#define VEC4_OPERATOR_BODY_VEC(op, sep) x op rhs.x sep y op rhs.y sep z op rhs.z sep w op rhs.w sep
+
+#define VEC4_OPERATOR_BODY_FLOAT(op, sep) x op rhs sep y op rhs sep z op rhs sep w op rhs sep
+
+#define VEC4_OPERATOR_BODY_VEC3(op, sep) x op rhs.x sep y op rhs.y sep z op rhs.z sep w sep
+
+#define VEC4_OPERATOR_BODY_VEC2(op, sep) x op rhs.x sep y op rhs.y sep z sep w sep
+
+namespace Siege
+{
+
+/**
+ * Struct representing a 3 dimensional vector.
+ */
+struct Vec4
+{
+    // Public constants
+
+    /**
+     * A vector in which all elements are 0.
+     */
+    static const Vec4 Zero;
+
+    /**
+     * A vector in which all elements are instantiated to 1.
+     */
+    static const Vec4 One;
+
+    /**
+     * A vector in which all elements except for the y value are 0.
+     */
+    static const Vec4 Up;
+
+    // 'Structors
+
+    /**
+     * Base Vector constructor.
+     * @param x the value to be given to the x dimension (default is zero)
+     * @param y the value to be given to the y dimension (default is zero)
+     * @param z the value to be given to the z dimension (default is zero)
+     * @param 2 the value to be given to the 2 dimension (default is zero)
+     */
+    Vec4(const float& x = 0.f, const float& y = 0.f, const float& z = 0.f, const float& w = 0.f) :
```

Comment:
Again, this is necessary because there are a few constructors in the class. Initialiser list constructors are only possible if no other constructors are available.